article-analysis-project/article-processing.py

```diff
@@ -0,0 +1,207 @@
+    for ent in docx.ents:
+        if ent.label_ == 'PERSON':
+            authors.append(ent.text)
+    return authors
+    
+def write_to_json(title, authors, gender_of_author, sources, count_male, count_female):
+     #Write to JSON
+        # Data to be written
+        dictionary ={ 
+            "title" : article.title, 
+            "Author" : authors, 
+            "Gender of Author": gender_of_author,
+            "Sources": np.array_str(sources),
+            "# of male sources" : count_male, 
+            "# of female sources" : count_female
+        } 
+    
+        # Serializing json  
+        json_object = json.dumps(dictionary, indent = 4) 
+    
+        # Writing to sample.json 
+        with open("example.json", "w") as outfile: 
+            outfile.write(json_object) 
+
+def extract_sources_displacy(sentences):
+    nlp = spacy.load('en_core_web_sm') #Load english
+    sources = []
+    for i, sentence in enumerate(sentences):
+        quote_words_bool = ("said" or "according to" or "saying" or "say" or "says") in sentence
+        sen = nlp(sentence)
+        if quote_words_bool:
+            #Merge noun chunks
+            #for noun_phrase in list(sen.noun_chunks):
+                #noun_phrase.merge(noun_phrase.root.tag_,noun_phrase.root.lemma_, noun_phrase.root.ent_type_)
+            #Find verb signifier
+            for token in sen:
+                if token.text == "said" or "according to " or "saying" or "say" or "says":
+                    for a in token.children:
+                        if a.dep_ == 'nsubj' and a.pos_ == 'PROPN':
+                            sources.append(a.text)
+    sources = np.array(sources)
+    sources = np.unique(sources) #Remove duplicates
+
+    #Remove non-names
+    for name in sources[:]:
+        if name.islower() or any(p in name for p in string.punctuation):
+            sources = np.delete(sources, np.where(sources == name))
+
+    #Find full names
+    sources_full_name = []
+    for source in sources:      
+        for sent in sentences:
+            split_sent = re.findall(r"[\w']+|[.,!?;]", sent)
+            if source in split_sent:
+                #split_sent = sent.split() #Split sentence into list
+                #index = split_sent.index(source)
+                #sources_full_name.append(split_sent[index - 1] + " " + split_sent[index])
+                docx = nlp(sent)
+                for ent in docx.ents:
+                    if ent.label_ == 'PERSON' and source in ent.text:
+                        sources_full_name.append(ent.text)
+                break
+    #sources_full_name = np.array(sources_full_name)
+    return sources_full_name 
+           
+def source_gender(sources, sentences):
+    nlp = spacy.load('en_core_web_sm')
+    genderize = Genderize(
+            user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36',
+            api_key='66982c609ec00aba67eee6fb146f6210',
```

Comment:
API key (also just to note that this code block has been repeated three times; might be worth creating a "generateGenderizeAgent" method somewhere.